README.md
# AceCRC

[![AUnit Tests](https://github.com/bxparks/AceCRC/actions/workflows/aunit_tests.yml/badge.svg)](https://github.com/bxparks/AceCRC/actions/workflows/aunit_tests.yml)

This library contains a number of CRC algorithms that were generated from
`pycrc` (https://pycrc.org) and programmatically converted to Arduino format to
use C++ namespaces to avoid name collisions and `PROGMEM` flash memory for
lookup tables to save static RAM. From the [list of pycrc supported
algorithms](https://pycrc.org/models.html), this library supports:

* CRC-8
* CRC-16-CCITT
* CRC-16-MODBUS
* CRC-32

For each algorithm, 4 different implementations were generated:

* bit: bit-by-bit
    * brute-force loops to calculate the polynomial divisions
    * smallest code size, but slowest
* nibble: 4-bit lookup table in flash memory (PROGMEM)
    * generates a lookup table of 16 elements
    * larger code size, but faster
* nibblem: 4-bit lookup table in static memory
    * same as nibble, but using static memory for lookup table
    * the "m" stands for "static memory" as opposed to "flash memory"
    * *2-7%* faster than `nibble` on AVR
    * **1.9X-2.7X** faster than `nibble` on ESP8266
    * no difference for all other processors
* byte: 8-bit lookup table in flash memory
    * generates a lookup table of 256 elements
    * largest code size, but fastest

The `pycrc` program generates `*.h` and `*.c` files containing C99 code with the
following definitions:

* `crc_t crc_init(void);`
* `crc_t crc_update(crc_t crc, const void *data, size_t data_len);`
* `crc_t crc_finalize(crc_t crc);`

This library converts the C99 code in the following way:

* each algorithm and variant is wrapped its own C++ namespace to avoid name
  collision
    * `ace_crc::crc8_bit`
    * `ace_crc::crc8_nibble`
    * `ace_crc::crc8_nibblem`
    * `ace_crc::crc8_byte`
    * `ace_crc::crc16ccitt_bit`
    * `ace_crc::crc16ccitt_nibble`
    * `ace_crc::crc16ccitt_nibblem`
    * `ace_crc::crc16ccitt_byte`
    * `ace_crc::crc16modbus_bit`
    * `ace_crc::crc16modbus_nibble`
    * `ace_crc::crc16modbus_nibblem`
    * `ace_crc::crc16modbus_byte`
    * `ace_crc::crc32_bit`
    * `ace_crc::crc32_nibble`
    * `ace_crc::crc32_nibblem`
    * `ace_crc::crc32_byte`
* a convenience function `crc_t crc_calculate(const void *data, size_t
  data_len)` is inserted into the header file of each namespace
    * calculates the CRC in one-shot
* the `crc_table` lookup table is moved into flash memory using `PROGMEM`
    * the static RAM usage of all CRC routines becomes zero (other than a few
      stack variables)
    * (Note that on platforms that do not support flash memory storage of
      data, `PROGMEM` is a no-op, and the various `pgm_read_{xxx}()` routines
      become just normal memory accessors.)
* the `static` keyword is removed from header files
    * not needed in C++
    * it was preventing the generation of doxygen docs for those functions
* the `#define CRC_ALGO_{XXX}` macro is converted into a `const uint8_t`
    * becomes part of its enclosing namespace, preventing name collision
* the typedef for `crc_t` is changed from `uint_fast8_t`, `uint_fast16_t`, and
  `uint_fast32_t` to `uint8_t`, `uint16_t`, and `uint32_t` respectively
    * affects only 32-bit processors, and only the `crc8`, `crc16ccitt`, and
      `crc16modbus` algorithms
    * see section [Integer Sizes](#IntegerSizes) below for more information

**TL;DR**: Use `crc32_nibble` in most cases, except on ESP8266 where you should
use `crc32_nibblem`.

**Version**: 1.1.1 (2023-07-12)

**Changelog**: [CHANGELOG.md](CHANGELOG.md)

## Table of Contents

* [HelloCRC](#HelloCRC)
* [Installation](#Installation)
    * [Dependencies](#Dependencies)
* [Documentation](#Documentation)
* [Usage](#Usage)
    * [Headers and Namespaces](#Headers)
    * [Core CRC Functions](#CoreFunctions)
    * [Integer Sizes](#IntegerSizes)
* [Resource Consumption](#ResourceConsumption)
    * [Memory Benchmarks](#MemoryBenchmarks)
    * [CPU Benchmarks](#CpuBenchmarks)
* [Recommendations](#Recommendations)
* [Background and Motivation](#Motivation)
    * [Other CRC Libraries](#OtherLibraries)
* [Bugs and Limitations](#Bugs)
* [System Requirements](#SystemRequirements)
    * [Hardware](#Hardware)
    * [Tool Chain](#ToolChain)
    * [Operating System](#OperatingSystem)
* [License](#License)
* [Feedback and Support](#FeedbackAndSupport)
* [Authors](#Authors)


<a name="HelloCRC"></a>
## HelloCRC

Here is the sample program from [examples/HelloCRC](examples/HelloCRC) that uses
the CRC-16-CCITT algorithm using a 4-bit lookup table (16 elements):

```C++
#include <Arduino.h>
#include <AceCRC.h>

using namespace ace_crc::crc16ccitt_nibble;

static const char CHECK_STRING[] = "123456789";
static const size_t LENGTH = sizeof(CHECK_STRING) - 1;

void setup() {
  Serial.begin(115200);
  while (!Serial); // Wait - Leonardo/Micro

  crc_t crc = crc_init();
  crc = crc_update(crc, CHECK_STRING, LENGTH);
  crc = crc_finalize(crc);
  Serial.print("0x");
  Serial.println((unsigned long) crc, 16);

  crc = crc_calculate(CHECK_STRING, LENGTH);
  Serial.print("0x");
  Serial.println((unsigned long) crc, 16);
}

void loop() {
}
```

This prints the hexadecimal numbers
```
0xE5CC
0xE5CC
```
as expected.

<a name="Installation"></a>
## Installation

The latest stable release is available in the Arduino IDE Library Manager.
Search for "AceCRC" and click install.

The development version can be installed by cloning the git repo:

* AceCRC (https://github.com/bxparks/AceCRC)

You can copy over the contents to the `./libraries` directory used by the
Arduino IDE. (The result is a directory named `./libraries/AceCRC`). Or you can
create symlinks from `./libraries` to these directories.

The `develop` branch contains the latest development.
The `master` branch contains the stable releases.

<a name="Dependencies"></a>
### Dependencies

This library has no external dependencies for client use. But the following
dependencies are required for development and testing purposes:

* To regenerate and rebuild the source code, you need pycrc
  (https://pycrc.org/).
* To run the unit tests under `tests`, you need AUnit
  (https://github.com/bxparks/AUnit)
* To run the unit tests under Linux or MacOS, you need EpoxyDuino
  (https://github.com/bxparks/EpoxyDuino)

<a name="Documentation"></a>
## Documentation

* [README.md](README.md) - this file
* [Doxygen docs](https://bxparks.github.io/AceCRC/html) hosted on GitHub Pages
* [examples/benchmarks](examples/benchmarks) - for memory and CPU consumption
  number for each algorithm on various microcontrollers.

<a name="Usage"></a>
## Usage

<a name="Headers"></a>
### Headers and Namespaces

Only a single header file `AceCRC.h` is required to use this library.

```C++
#include <AceCRC.h>
```

Then select the namespace of the algorithm that you want to use. For example,
to use the version of `crc32` which uses a 4-bit table (16 entries), which
are stored in static memory instead of flash memory, use the following:

```C++
using namespace ace_crc::crc32_nibblem;
```

To use the 8-bit table, which are stored in flash memory, use:

```C++
using namespace ace_crc::crc32_byte;
```

<a name="CoreFunctions"></a>
### Core CRC Functions

All algorithms and their variants are placed in separate C++ namespaces so they
do not collide, and you can use multiple CRC algorithms in a single program
without worrying about name collision.

The function names and their signatures from the underlying generated code from
`pycrc` remain unchanged. For reference, the principle functions are:

* `crc_t crc_init(void);`
* `crc_t crc_update(crc_t crc, const void *data, size_t data_len);`
* `crc_t crc_finalize(crc_t crc);`

See the [examples/HelloCRC](examples/HelloCRC) example code to see how these
functions are used. The `crc_update()` function can be called multiple times
with additional data, before calling `crc_finalize()`.

This library adds the following convenience function to each header file in each
namespace:

* `crc_t crc_calculate(const void *data, size_t data_len);`

This function replaces the three separate calls to `crc_init()`, `crc_update()`,
`crc_finalize()` with a single call.

<a name="IntegerSizes"></a>
### Integer Sizes

By default, the `pycrc` program generates C99 code which contains one of the
following definitions of the `crc_t` type:

```C++
typedef uint_fast8_t crc_t;
typedef uint_fast16_t crc_t;
typedef uint_fast32_t crc_t;
```

These are converted by this library to these instead:
```C++
typedef uint8_t crc_t;
typedef uint16_t crc_t;
typedef uint32_t crc_t;
```

On 8-bit processors, `uint_fast8_t` is identical to `uint8_t`, `uint_fast16_t`
is identical to `uint16_t`, and `uint_fast32_t` is identical to `uint32_t`.

On 32-bit processors (e.g. SAMD, ESP8266, ESP32), `uint_fast8_t` and
`uint_fast16_t` are both defined to be `uint32_t`, presumably because the 32-bit
integer type is faster for most operations (but not always). The main effect of
these definitions is to increase the size of the `crc_table` for the CRC-8 and
CRC-16-CCITT algorithms by a factor of 4x or 2x, compared to what they could be.

After collecting the CPU and memory consumption results of various algorithms on
different microcontrollers in [examples/benchmarks](examples/benchmarks), I
found that using the smaller `uint8_t` or `uint16_t` did not affect the speed of
the algorithms very much (2-14%). Some were got slightly slower, but some
actually got slightly faster using the exact `uint8_t` and `uint16_t` types,
which are supposed to be slower.

The bigger difference is the sizes of the internal `crc_table` which become a
lot smaller when the `crc_t` becomes smaller. It also seemed potentially
confusing for the end user if `sizeof(crc_t)` returned 4 instead of 1 or 2 when
using the CRC-8 or CRC-16-CCITT algorithms.

I concluded that it was better in the Arduino microcontroller environments to
make the `crc_t` type correspond to the exact sized integer types (`uint8_t`,
`uint16_t`, `uint32_t`).

The `size_t` is used as the type of the `data_len` parameter of the
`crc_update()` or the `crc_calculate()` function. On 8-bit processors `size_t`
is 2 bytes, and on 32-bit processors `size_t` is 4 bytes. To calculate the CRC
of an array that's longer than 64 kiB long on 8-bit processors, it must be
broken down into chunks smaller than 64 kiB and processed through the
`crc_update()` function multiple times. I suspect that this situation will
happen only rarely on 8-bit processors which tend to deal with fairly small
amounts of data.

<a name="ResourceConsumption"></a>
## Resource Consumption

<a name="MemoryBenchmarks"></a>
### Memory Benchmarks

I wrote a bunch of scripts in
[examples/benchmarks/MemoryBenchmark](examples/benchmarks/MemoryBenchmark) to
automatically gather the flash and static RAM consumption of various CRC
algorithms on various microcontrollers. The results are summarized in the
`README.md` in that directory. All algorithms (except for `xxx_nibblem`
variants) consume no static RAM, because their lookup tables are located in
flash using `PROGMEM`.

Here are rough flash memory consumption for each algorithm:

* `crc8_bit`: 64-130 bytes
* `crc8_nibble`: 80-150 bytes
* `crc8_nibblem`: 80-150 bytes
* `crc8_byte`: 290-360 bytes
* `crc16ccitt_bit`: 80-140 bytes
* `crc16ccitt_nibble`: 100-190 bytes
* `crc16ccitt_nibblem`: 100-190 bytes
* `crc16ccitt_byte`: 560-630 bytes
* `crc16modbus_bit`: 120-200 bytes
* `crc16modbus_nibble`: 100-180 bytes
* `crc16modbus_nibblem`: 100-180 bytes
* `crc16modbus_byte`: 560-630 bytes
* `crc32_bit`: 110-180 bytes
* `crc32_nibble`: 140-210 bytes
* `crc32_nibblem`: 140-210 bytes
* `crc32_byte`: 1080-1140 bytes

<a name="CpuBenchmarks"></a>
### CPU Benchmarks

The CPU performance of each CRC algorithm and variant is given in
[examples/benchmarks/CpuBenchmark](examples/benchmarks/CpuBenchmark) as units of
microseconds per kiB (1024 bytes).

For 8-bit processors (e.g. Nano, Micro), the numbers are roughly:
* "bit": 7800-18000 micros/kiB
* "nibble": 5300-7600 micros/kiB
* "nibblem": 4900-7200 micros/kiB
* "byte": 900-2200 micros/kiB

For 32-bit processors (e.g. SAMD, ESP8266, ESP32), the numbers are in the range
of:
* "bit": 400-2800 micros/kiB
* "nibble": 110-700 micros/kiB
* "nibblem": 110-700 micros/kiB
* "byte": 50-400 micros/kiB

<a name="Recommendations"></a>
## Recommendations

**TL;DR**: Use `crc32_nibble` in most cases, except on ESP8266 where you should
use `crc32_nibblem`.

The benchmark numbers from `CpuBenchmark` and `MemoryBenchmark` are combined
into a single place in [examples/benchmarks](examples/benchmarks) for
convenience. Here are 2 samples:

**Arduino Nano**

```
+--------------------------------------------------------------+
| CRC algorithm                   |  flash/  ram |  micros/kiB |
|---------------------------------+--------------+-------------|
| crc8_bit                        |     64/    0 |        7808 |
| crc8_nibble                     |    130/    0 |        7356 |
| crc8_nibblem                    |    134/   16 |        7224 |
| crc8_byte                       |    292/    0 |         916 |
|---------------------------------+--------------+-------------|
| crc16ccitt_bit                  |     84/    0 |       11888 |
| crc16ccitt_nibble               |    134/    0 |        5296 |
| crc16ccitt_nibblem              |    138/   32 |        5040 |
| crc16ccitt_byte                 |    562/    0 |        1496 |
|---------------------------------+--------------+-------------|
| crc16modbus_bit                 |    120/    0 |       11692 |
| crc16modbus_nibble              |    132/    0 |        5232 |
| crc16modbus_nibblem             |    134/   32 |        4912 |
| crc16modbus_byte                |    562/    0 |        1496 |
|---------------------------------+--------------+-------------|
| crc32_bit                       |    188/    0 |       18248 |
| crc32_nibble                    |    204/    0 |        7616 |
| crc32_nibblem                   |    208/   64 |        7104 |
| crc32_byte                      |   1106/    0 |        2272 |
|---------------------------------+--------------+-------------|
| CRC32                           |    208/    0 |        7688 |
| Arduino_CRC32                   |   1112/ 1024 |        2144 |
| FastCRC                         |   4262/    0 |        2160 |
+--------------------------------------------------------------+
```

**ESP8266**

```
+--------------------------------------------------------------+
| CRC algorithm                   |  flash/  ram |  micros/kiB |
|---------------------------------+--------------+-------------|
| crc8_bit                        |    128/    0 |        1513 |
| crc8_nibble                     |    144/    0 |         488 |
| crc8_nibblem                    |    112/   16 |         270 |
| crc8_byte                       |    336/    0 |         246 |
|---------------------------------+--------------+-------------|
| crc16ccitt_bit                  |    128/    0 |        1820 |
| crc16ccitt_nibble               |    160/    0 |         669 |
| crc16ccitt_nibblem              |    144/   32 |         283 |
| crc16ccitt_byte                 |    608/    0 |         375 |
|---------------------------------+--------------+-------------|
| crc16modbus_bit                 |    160/    0 |        1937 |
| crc16modbus_nibble              |    160/    0 |         642 |
| crc16modbus_nibblem             |    128/   32 |         244 |
| crc16modbus_byte                |    608/    0 |         375 |
|---------------------------------+--------------+-------------|
| crc32_bit                       |    144/    0 |        1567 |
| crc32_nibble                    |    176/    0 |         591 |
| crc32_nibblem                   |    160/   64 |         244 |
| crc32_byte                      |   1104/    0 |         340 |
|---------------------------------+--------------+-------------|
| CRC32                           |    224/    0 |         834 |
| Arduino_CRC32                   |   1120/ 1024 |         155 |
| FastCRC                         |   4656/    0 |         458 |
+--------------------------------------------------------------+
```

Comparing the different variants ("bit", "nibble" and "byte"), it seems that the
"nibble" variants (4-bit lookup table) offer a good tradeoff between
flash memory consumption and CPU speed in the following ways:

* Compared to the "bit" versions, the "nibble" variants are about the same size
  but they can be up to ~2X (8-bit) to ~5X (32-bit) faster.
* Compared to the "byte" versions, the "nibble" variants can be 4-5X
  smaller in flash size, but only about 3-4X (8-bit) to 1.5-2X (32-bit) slower.

The CRC-8 algorithm has the unfortunate property that arrays of zeros of
different lengths (e.g. 1 zero or 2 zeros) have the exact same CRC (0). The
other two (CRC-16-CCITT and CRC32) are able to distinguish strings of zeroes of
different lengths. In terms of flash size and performance, the CRC-8 algorithm
is not all that much faster than the CRC-16-CCITT, even on 8-bit processors. For
these reasons, the CRC-8 algorithm is not recommended, unless you are really
strapped for flash bytes.

The CRC-16-CCITT was selected for this library over the CRC-16 algorithm because
CRC-16 suffers the same problem as CRC-8 with regards to arrays of zeros of
different lengths. The CRC-16-CCITT uses a non-zero XOR-input which allows it to
distinguish different lengths of zeros.

Between the CRC-16-CCITT and CRC-32 algorithms, if we look at the `_nibble`
variants, there is very little difference in flash size and CPU speed, even on
8-bit processors. On 32-bit processors, the CRC-32 is actually faster. The
advantage of CRC-32 over CRC-16-CCITT is that 32 bits will be able to detect
longer sequences of errors than 16 bits.

The `nibblem` variant is the same as the `nibble` variant on all processors
except for AVR (small speed improvement) and ESP8266 (1.9X-2.7X speed
improvement). The cost is that the 16-64 bytes of static memory is consumed to
store the lookup table, instead of keeping them in flash memory.

Putting all these together, here are my recommended algorithms in decreasing
order of preference:
1. `crc32_nibble` in most situtations for a balance of flash size (~200 bytes)
   and speed
    * Except on ESP8266, where `crc32_nibblem` should be used to gain a
      factor of **2.7X** in performance, at a cost of only 64 byes of static RAM
      (out of a maximum size of 80kB).
2. `crc16ccitt_nibble` to save 60 bytes of flash on 8-bit AVR processors, with
   a reasonable amount of error detection
    * Use `crc16ccitt_nibblem` if you want a 5% performance increase on an AVR,
      at a cost of 32 bytes of static memory. In most cases, this is probably
      not worth it.
3. `crc32_byte` if you need the fastest algorithm and you have 1100 bytes of
   flash to spare
    * Except on ESP8266, where `crc32_nibblem` is even faster than `crc32_byte`,
      while consuming 1 kB of less flash memory, and costing only 64 bytes of
      extra static memory.
4. `crc16ccitt_bit` if you need to implement a very tiny CRC algorithm (90 bytes
   on an 8-bit AVR processor), and you are not worried about speed
5. `crc8_bit` if you need to implement the absolute smallest CRC algorithm (80
   bytes on an 8-bit AVR processor), and you are not worried about speed, and
   you can tolerate high chances of corruption

On the ESP8266, the `crc32_nibblem` is the hands-down winner. It is faster than
all other algorithms, while consuming only 160 bytes of flash and 64 bytes of
static ram.

You can consult the results in [examples/benchmarks](examples/benchmarks) to
determine exactly how you want to make the space versus time tradeoff for your
specific application.

<a name="Motivation"></a>
## Background and Motiviation

Before writing this library, I did not understand how CRC algorithms worked and
how they were implemented. I just knew that they calculated a checksum. I had
been using the `crc32()` algorithm from the FastCRC library
(https://github.com/FrankBoesing/FastCRC) but when I dug in a little deeper, I
discovered that it was configured to use "large tables" by default, which is
why it is "fast". But that meant that the `crc32()` algorithm was pulling in a
table of 1024 elements of 4 bytes each, for a total of 4kB.

The `crc32()` function was being called from my `CrcEeprom` class
(https://github.com/bxparks/AceUtils), which I had included in various
applications running on an Arduino Nano or Sparkfun Pro Micro with only 32kB of
flash. Due to my ignorance, I was using at least 1/8 of my flash memory budget
just to calculate the CRC32!

I decided to figure out how to write my own CRC routines. The best reference I
found was [A Painless Guide to CRC Error Detection
Algorithms](http://ross.net/crc/download/crc_v3.txt) by Ross Williams written in
1993. Once I read that, I could understand how a code generator such as `pycrc`
(https://pycrc.org/) worked. For each algorithm type (with a given polynomial

CHANGELOG.md
# Changelog

* Unreleased
* 1.1.1 (2023-07-12)
    * Update supported boards and tiers
        * Add SAMD21 and SAMD51 boards to Tier 1
            * Add 2 SAMD boards from 2 different companies, to test their
              Arduino Core software:
                * Seeeduino XIAO M0 (SAMD21 48MHz ARM Cortex-M0+)
                * Adafruit ItsyBitsy M4 (SAMD51 120MHz ARM Cortex-M4)
            * SAMD21 and SAMD51 boards are back in Tier 1, as long as they use
              the traditional Arduino API instead of the new
              [ArduinoCore-API](https://github.com/arduino/ArduinoCore-api).
            * Fortunately most third party SAMD21 and SAMD51 boards continue to
              use the traditional Arduino API.
        * Move Teensy 3.2 to Tier 2 ("Should work but not tested often")
            * This board is entering end-of-life.
            * As well, the Teensyduino environment integrates with the Arduino
              IDE and CLI in a way that's different than all other third-party
              Arduino boards. Some of my automation scripts do not work with
              Teensyduino, so it becomes very time consuming to test the Teensy
              boards.
            * All Teensy boards are now in Tier 2.
    * This is a maintenance release, no changes to code.
* 1.1.0 (2023-02-03)
    * Upgrade tool chain
        * Arduino IDE from 1.8.13 to 1.8.19
        * Arduino CLI from 0.15.2 to 0.27.1
        * Arduino AVR Core from 1.8.3 to 1.8.5
        * SparkFun AVR Core 1.8.13
        * SparkFun SAMD Core from 1.8.1 to 1.8.7
        * STM32duino Core from 1.9.0 to 2.3.0
        * ESP8266 Core from 2.7.4 to 3.0.2
        * ESP32 Core from 1.0.6 to 2.0.5
        * TeensyDuino from 1.54 to 1.57
    * Remove SAMD21 boards.
        * No longer supported.
    * Add CRC-16-MODBUS from pycrc (see
      [Issue#12](https://github.com/bxparks/AceCRC/issues/12)).
* 1.0.1 (2021-04-19)
    * Update EpoxyDuino Makefiles for better compatibility with FreeBSD.
    * Add recommendation to use `nibblem` on ESP8266.
    * Upgrade to ESP32 Arduino Core v1.0.6 and update benchmark numbers (no
      significant changes).
    * No functional change in this release.
* 1.0 (2021-02-21)
    * Add `nibblem` variant.
        * Same as `nibble` but the 16-element lookup table is stored in static
          memory instead of flash memory.
        * On ESP8266, this causes a **1.9X-2.7X** speed improvement because
          accessing the flash memory access is much slower than static memory.
        * Only a 2-7% speed improvement for AVR processors.
    * Graduate to v1.0.
* 0.4.2 (2021-01-22)
    * Update UnixHostDuino 0.4 to EpoxyDuino 0.5
    * No functional change in this release.
* 0.4.1 (2021-01-19)
    * Add official support for STM32 by validating against STM32 Blue Pill
      board. Update CpuBenchmark and MemoryBenchmark numbers.
    * Re-add benchmark numbers for Teensy 3.2. Able to upload to it again
      after reinstalling Ubuntu 20.04.
    * No functional change in this release.
* 0.4 (2020-12-10)
    * Convert internal loop or index variable inside `crc_update()` fgrom
      `unsigned int` to a one-byte `uint8_t`. Improves CPU performance by 6%-34%
      on 8-bit processors. See
      [examples/benchmarks/CpuBenchmark](examples/benchmarks/CpuBenchmark). No
      change in performance for 32-bit processors.
    * Add MemoryBenchmark numbers for the same third-party CRC libraries as
      CpuBenchmark.
* 0.3.2 (2020-12-09)
    * Add CpuBenchmark numbers for some third-party CRC libraries for curiosity.
    * Rewrite README.md for clarity.
    * Use Arduino-CLI to gather MemoryBenchmark results faster.
    * No functional change.
* 0.3.1 (2020-12-03)
    * Use exact sized `uint16_t` and `uint32_t` as the typedef for `crc_t`,
      instead of the default `uint_fast16_t` and `uint_fast32_t` produced by
      pycrc. This change affects only the 32-bit processors where
      `uint_fast16_t` is defined to be 32 bits instead of 16 bits. Some of the
      `crc16ccitt` algorithms become slightly slower but some become slightly
      faster. The biggest difference is that the `crc_table` sizes for the
      `cc16ccitt` algorithms all become 1/2 the size, which is a big win.
    * Add CRC-8 algorithms in `crc8_bit`, `crc8_nibble`, and `crc8_byte`
      namespaces. Update MemoryBenchmark and CpuBenchmark numbers.
    * Add recommendation list of various algorithms in README.md.
* 0.3 (2020-12-03)
    * Botched release, deleted.
* 0.2 (2020-12-03)
    * Generate CPU times of each CRC algorithm and variant using
      [examples/benchmarks/CpuBenchmark](examples/benchmarks/CpuBenchmark).
    * Create combined CPU and Memory benchmark tables in
      [examples/benchmarks](examples/benchmarks).
    * Insert a `crc_calculate()` convenience function into the header files so
      that the CRC can be calculated in one-shot, instead of forcing the user to
      call `crc_init()`, `crc_update()` and `crc_finalize()`.
    * Remove `static` keyword from `*.hpp` files. Not needed in C++ and
      prevents doxygen doc generation of those functions.
* 0.1.1 (2020-11-29)
    * Fix embarrassing typos in library.properties.
* 0.1 (2020-11-29)
    * Move `crc_table` lookup table to flash using `PROGMEM`.
    * Add `MemoryBenchmark` to determine flash and static RAM consumption of
      each algorithm.
    * Add doxygen docs.
* Initial Commit (2020-11-29)
    * Initial commit with 2 CRC algorithms: `crc32()` and `crc16ccitt()`


tools/generate.sh
#!/bin/bash
#
# Generate CRC source code using pycrc (https://pycrc.org). Then convert the
# C-code into Arduino-compatible C++ code using the old Unix ed(1) editor.
#
# The 4 supported algorithms from pycrc are:
#
#   * bit: --algorithm bit-by-bit-fast
#   * nibble: --algorithm table-driven --table-idx-width 4
#   * nibblemem: --algorithm table-driven --table-idx-width 4 (with lookup
#     table in static memory instead of flash memory)
#   * byte: --algorithm table-driven --table-idx-width 8

set -eu

# Set this to 0 to prevent conversion to PROGMEM and <pgmspace.h>.
# Might be useful to turn it off for debugging.
CONVERT_TO_PROGMEM=1

# Set this to 1 to preserve a copy of the original *.c and *.h files from pycrc.
# Useful for debugging purposes.
PRESERVE_C_FILES=0

function usage() {
    echo "Usage: generate.sh [--help|-h] \
    --model {crc-8|crc-16-ccitt|crc-16-modbus|crc-32} \
    --algotag {bit|nibble|nibblem|byte} \
    {--header | --source}"
    exit 1
}

# Generate the *.h header file.
function generate_h() {
    local modelroot=$1
    local model=$2
    local flags="$3"
    local fileroot="${modelroot}_${algotag}"

    python3 -m pycrc \
        --model $model \
        $flags \
        --generate h \
        -o $fileroot.h
}

# Generate the *.c source file
function generate_c() {
    local modelroot=$1
    local model=$2
    local flags="$3"
    local fileroot="${modelroot}_${algotag}"

    python3 -m pycrc \
        --model $model \
        $flags \
        --generate c \
        -o $fileroot.c
}

# Convert C-style *.h file to Arduino C++ *.hpp file
function convert_h_to_hpp() {
    local modelroot=$1
    local algotag=$2
    local model=$3
    local old_header_guard=$4
    local new_header_guard=$5
    local fileroot="${modelroot}_${algotag}"

    ed $fileroot.h > /dev/null <<EOF
/ \*\//
i
 *
 * Auto converted to Arduino C++ on $(date)
 * by AceCRC (https://github.com/bxparks/AceCRC).
 * DO NOT EDIT
.
/#ifndef $old_header_guard/
.,+1s/$old_header_guard/$new_header_guard/
/#ifdef __cplusplus/
.,+2c
namespace ace_crc {
namespace $fileroot {
.
/#define CRC_ALGO/
s/#define/static const uint8_t/
s/ \([0-9][0-9]*\)$/ = \1;/
/#ifdef __cplusplus/
.,+2c
/**
 * Calculate the crc in one-shot.
 * This is a convenience function added by AceCRC.
 *
 * \param[in] data     Pointer to a buffer of \a data_len bytes.
 * \param[in] data_len Number of bytes in the \a data buffer.
 */
inline crc_t crc_calculate(const void *data, size_t data_len) {
  crc_t crc = crc_init();
  crc = crc_update(crc, data, data_len);
  return crc_finalize(crc);
}

} // $fileroot
} // ace_crc
.
/#endif.*${old_header_guard}/
s/$old_header_guard/$new_header_guard/
1,\$s/^static //
/typedef uint_fast/s//typedef uint/
w $fileroot.hpp
q
EOF

    if [[ $PRESERVE_C_FILES == 0 ]]; then
        rm $fileroot.h
    fi
}

# Convert C-style *.c file to Arduino C++ *.cpp file
function convert_c_to_cpp() {
    local modelroot=$1
    local algotag=$2
    local model=$3
    local pgm_read_func=$4
    local fileroot="${modelroot}_${algotag}"

    # Determine the name of the loop or index variable to convert from
    # 'unsigned in' to 'uint8_t'.
    local loop_or_index_variable
    case $algotag in
        bit) loop_or_index_variable='i' ;;
        nibble|nibblem|byte) loop_or_index_variable='tbl_idx' ;;
    esac

    # The bit-by-bit algorithm does not use a crc_table, so we cannot attempt to
    # convert it to PROGMEM, so needs a separate ed(1) script. Otherwise ed(1)
    # script fails with error.
    if [[ $algotag == 'bit' ||
            $algotag == 'nibblem' ||
            $CONVERT_TO_PROGMEM == 0 ]]; then
        ed $fileroot.c > /dev/null <<EOF
/ \*\//
i
 *
 * Auto converted to Arduino C++ on $(date)
 * by AceCRC (https://github.com/bxparks/AceCRC).
 * DO NOT EDIT
.
/#include "$fileroot.h"/
c
#include "$fileroot.hpp" // header file converted by AceCRC
.
/^$/a
namespace ace_crc {
namespace $fileroot {

.
/crc_update/
/unsigned int $loop_or_index_variable;/c
    uint8_t $loop_or_index_variable;
.
\$a

} // $fileroot
} // ace_crc
.
w $fileroot.cpp
q
EOF

    # These algorithms use the 'crc_table' so move that table into flash memory
    # using PROGMEM, then convert the crc_table[i] expressions with the
    # equivalent ones using pgm_read_word() or pgm_read_dword().
    else
        ed $fileroot.c > /dev/null <<EOF
/ \*\//
i
 *
 * Auto converted to Arduino C++ on $(date)
 * by AceCRC (https://github.com/bxparks/AceCRC).
 * DO NOT EDIT
.
/#include "$fileroot.h"/
c
#include "$fileroot.hpp" // header file converted by AceCRC
.
i
#if defined(ARDUINO_ARCH_AVR) || defined(ARDUINO_ARCH_SAMD)
  #include <avr/pgmspace.h>
#else
  #include <pgmspace.h>
#endif
.
/^$/a
namespace ace_crc {
namespace $fileroot {

.
/crc_table/s/= {/PROGMEM &/
g/crc_table\[\(tbl_idx.*\)\]/s//($pgm_read_func(crc_table + (\1)))/
/crc_update/
/unsigned int $loop_or_index_variable;/c
    uint8_t $loop_or_index_variable;
.
\$a

} // $fileroot
} // ace_crc
.
w $fileroot.cpp
q
EOF
    fi

    if [[ $PRESERVE_C_FILES == 0 ]]; then
        rm $fileroot.c
    fi
}

header=0
source=0
model=''
algotag=''
while [[ $# -gt 0 ]]; do
    case $1 in
        --help|-h) usage ;;
        --header) header=1 ;;
        --source) source=1 ;;
        --model) shift; model="$1" ;;
        --algotag) shift; algotag="$1" ;;
        --) shift; break ;;
        -*) echo "Unknown flag '$1'" 1>&2; usage 1>&2 ;;
        *) break ;;
    esac
    shift
done

if [[ "$algotag" == '' ]]; then
    echo '--algotag is required'
    usage
fi
if [[ "$model" == '' ]]; then
    echo '--model is required'
    usage
fi

# Normalize names:
#   * "crc-16-ccitt" to "crc16ccitt"
#   * "crc-16-modbus" to "crc16modbus"
modelroot=$(echo $model | sed -e 's/-//g')
modelroot_upper="$(echo $modelroot | tr '[a-z]' '[A-Z]')"
algotag_upper="$(echo $algotag | tr '[a-z]' '[A-Z]')"

# CPP header guard e.g. "CRC16CCITT_BIT_H"
old_header_guard="${modelroot_upper}_${algotag_upper}_H"
new_header_guard="ACE_CRC_${modelroot_upper}_${algotag_upper}_HPP"

# Convert algotag to pycrc flags.
case $algotag in
    bit) pycrc_flags='--algorithm bit-by-bit-fast' ;;
    nibble) pycrc_flags='--algorithm table-driven --table-idx-width 4' ;;
    nibblem) pycrc_flags='--algorithm table-driven --table-idx-width 4' ;;
    byte) pycrc_flags='--algorithm table-driven --table-idx-width 8' ;;
    *) echo "Unknown --algotag '$algotag'"; usage ;;
esac

# Determine the <progmem.h> function that extracts the lookup table.
case $model in
    crc-8) pgm_read_func='pgm_read_byte' ;;
    crc-16-ccitt) pgm_read_func='pgm_read_word' ;;
    crc-16-modbus) pgm_read_func='pgm_read_word' ;;
    crc-32) pgm_read_func='pgm_read_dword' ;;
    *) echo 'Unknown --model '$model''; usage ;;
esac

# Perform the action that was requested by --header or --source.
if [[ $header == 1 ]]; then
    generate_h $modelroot $model "$pycrc_flags"
    convert_h_to_hpp $modelroot $algotag $model \
        $old_header_guard $new_header_guard
fi

if [[ $source == 1 ]]; then
    generate_c $modelroot $model "$pycrc_flags"
    convert_c_to_cpp $modelroot $algotag $model $pgm_read_func
fi
